--- src/engine/qsarray_object.cpp.orig	2006-12-31 03:12:09.000000000 +0100
+++ src/engine/qsarray_object.cpp	2006-12-31 03:13:49.000000000 +0100
@@ -138,6 +138,10 @@
 {
     QSObject v = *obj;
     int len = length( obj );
+    QSPropertyMap *propMap = properties( obj );
+    if (!propMap)
+        return QVariant();
+        
     if ( t == QVariant::StringList ) {
 	QStringList l;
 	for ( int i = 0; i < len; ++i )
@@ -146,9 +150,6 @@
 	return var;
     } else if ( t == QVariant::Map ) {
 	QMap<QString, QVariant> m;
-	QSPropertyMap *propMap = properties( obj );
-        if (!propMap)
-            return m;
 	for ( QSPropertyMap::ConstIterator it = propMap->begin();
 	      it != propMap->end(); ++it )
 	    m.insert( it.key(), (*it).object.toVariant( QVariant::Invalid ) );
@@ -161,9 +162,25 @@
 	return QVariant( l.join( QString::fromLatin1(",") ) );
     } else {
 	QValueList<QVariant> l;
-	for ( int i = 0; i < len; ++i )
-	    l << v.get( QString::number( i ) ).toVariant( QVariant::Invalid );
-	QVariant var = l;
+        QVariantMap m;
+        bool conversion_ok = true;
+        for ( QSPropertyMap::ConstIterator it = propMap->begin(); it != propMap->end(); ++it ) {
+            if (conversion_ok) {
+                int index = it.key().toInt(&conversion_ok);
+                if (conversion_ok) {
+                    // Pad to the right index
+                    while (l.size() <= index) l << QVariant();
+                    l[index] = (*it).object.toVariant(QVariant::Invalid);
+                }
+            }
+            m[it.key()] = (*it).object.toVariant(QVariant::Invalid);
+        }
+        QVariant var;
+        if (conversion_ok)
+            var = l;
+        else
+            var = m;	
+        
 	return var;
     }
 }
